### PR TITLE
[WiP] CHANGE-20414 Remove Journey

### PR DESCRIPTION
**Pull request by RyanMcDonald**
Removed the Journey gem which is now included in the ActionDispatch module in Rails 4.2. The 4.2 upgrade also brought a bunch of other code changes in ActionDispatch, so some refactoring was necessary.

### CODE REVIEW

**Comment by quaelin** [comment]
@mdimas What are your thoughts regarding trying to make this gem support *both* Rails 3.2 and Rails 4.2, rather than one or the other?  Originally, I was thinking that would give us the most flexibility... but I'm not sure how much it complicates the gem to try and support two different versions of Rails...

File: config/secrets.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+development:
+  secret_key_base: 8b16915e35590d84c65509821a16dd29561866c937e46d0cce6be9e815b60a0cea81ad96ed5800d56a5de9b31f9bf0a32a45d7435e7ffd91aa52bb185ce64555
+
+test:
+  secret_key_base: 53717d6a1d28dc695039f211666e221263891e60e5a19259c22d025f4a601c8ac3b3ed16ec5878863731adbe1dc6f7f1f4b5b4d006b79e19847d5ca44a2eb708
+
+# Do not keep production secrets in the repository,
+# instead read values from the environment.
+production:
```
Comment (RyanMcDonald) [inline config/secrets.yml]: How do we implement this for production? Rails 4.2 apparently requires these secret keys now? Seems weird we have to do this for a gem, though. For the development and test secrets I just did:
```
$ irb
>> require 'securerandom'
=> true
>> SecureRandom.hex(64)
=> "3fe397575565365108556c3e5549f139e8078a8ec8fd2675a83de96289b30550a266ac04488d7086322efbe573738e7b3ae005b2e3d9afd718aa337fa5e329cf"
>> exit
```

Comment (quaelin) [inline config/secrets.yml]: For the Ruby services, though they have been upgraded to Rails 4.2, we *don't* need `secrets.yml`, and I think it's because session storage is disabled:
https://github.com/change/petition_service/blob/a536d4793b8d62113fd1bfd36020268148f61749/config/application.rb#L96

For monorail we probably can't disable session storage, but for the gem perhaps we can...

File: spec/i18n_routable/i18n_routable_spec.rb [hunk 2]
```diff
@@ -160,44 +160,44 @@ def resolve(path, http_verb=:get)
...
-      I18n.locale = 'es'
-
-      route_hash = hash_for_post_url(:id => 1)
-      route_hash = route_hash.slice!(:use_route)
-      route_hash.should eql :id => 1, :action => "show", :controller => "posts", :only_path => false
-      url_for(route_hash).should eql "http://www.example.com/es/puestos/1"
-
-    end
-  end
+  # context 'hash_for helpers' do
```
Comment (RyanMcDonald) [inline spec/i18n_routable/i18n_routable_spec.rb]: The `hash_for` URL helper methods were removed in Rails 4.2. See [3.2](https://github.com/rails/rails/blob/3-2-stable/actionpack/lib/action_dispatch/routing/route_set.rb) vs [4.2](https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/routing/route_set.rb) where the `define_hash_access` method isn't in 4.2. Does this mean these tests won't be necessary anymore? I couldn't see a way of refactoring these tests to make them applicable.

File: spec/spec_routes.rb [hunk 3]
```diff
@@ -26,7 +26,7 @@ def draw!
 
         localize do
 
-          get :constraints => proc { false }, "*path" => "FooController#foo"
+          get :constraints => proc { false }, "*path" => "foocontroller#foo"
```
Comment (RyanMcDonald) [inline spec/spec_routes.rb]: This had to be changed because of this error in 4.2: `'FooController' is not a supported controller name. This can lead to potential routing problems. See http://guides.rubyonrails.org/routing.html#specifying-a-controller-to-use`

**Comment by mdimas** [comment]
@quaelin and @RyanMcDonald, I've seen both approaches taken for gems when dealing with Rails compatibility. I think I like the approach of making a different release for 4.x support. So if our main release branch is 0.0.4 I would suggest making a new 0.1.0 branch that only supports Rails 4.

File: spec/i18n_routable/i18n_routable_spec.rb [hunk 4]
```diff
@@ -84,9 +84,9 @@ def resolve(path, http_verb=:get)
       posts_path(locale: :fr, :i18n_posts => 'puestos').should eql '/fr/messages'
     end
 
-    it 'should work without passing parameters in a hash' do
-      post_path("post-title").should eql "/posts/post-title"
-    end
+    # it 'should work without passing parameters in a hash' do
+    #   post_path("post-title").should eql "/posts/post-title"
```
Comment (RyanMcDonald) [inline spec/i18n_routable/i18n_routable_spec.rb]: Still trying to get this test to pass. The culprit is on lib/i18n_routable/generating/localizable_url_helper.rb:15. The hash_for method was used previously, but with 4.2, it breaks if you don't pass in an options hash. Compare [3.2](https://github.com/rails/rails/blob/3-2-stable/actionpack/lib/action_dispatch/routing/route_set.rb#L209) vs [4.2](https://github.com/rails/rails/blob/21f7bcbaa7709ed072bb2e1273d25c09eeaa26d9/actionpack/lib/action_dispatch/routing/route_set.rb#L294) where it sets options to nil if there's no options hash, so if we try to set `options[:_positional_args] = args` it errors. I fixed this by setting options to an empty hash instead of nil, but then the test fails with
```
expected: "/posts/post-title"
            got: "?_positional_keys=%5B%3Aid%2C+%3Aformat%5D"
```

File: config/secrets.yml (hunk 1 above)
Comment (mdimas) [inline config/secrets.yml]: Hoes does the gem require sessions? Seems like that is outside of the scope of what the gem is responsible for.

File: i18n_routable.gemspec [hunk 5]
```diff
@@ -18,9 +18,10 @@ Gem::Specification.new do |s|
   s.executables   = `git ls-files -- bin/*`.split("\n").map{ |f| File.basename(f) }
   s.require_paths = ["lib"]
 
-  s.add_dependency "rails", '~> 3.2.0'
-  s.add_dependency 'i18n', '~> 0.6', '< 0.7'
+  s.add_dependency "rails", '~> 4.2.0'
+  s.add_dependency 'i18n', '~> 0.7'
```
Comment (mdimas) [inline i18n_routable.gemspec]: I think this is going to give us an issue with "en-US" which is an unsupported locale in i18n v0.7

File: lib/i18n_routable/config.rb [hunk 6]
```diff
@@ -5,6 +5,7 @@ module Config
     self.localize_config = {}
 
     def setup! options
+      I18n.config.enforce_available_locales = false
```
Comment (mdimas) [inline lib/i18n_routable/config.rb]: where do we get errors without this line?

File: lib/i18n_routable/translation_assistant.rb [hunk 7]
```diff
@@ -15,16 +15,18 @@ def untokenize_segment segment
...
+      new_path_spec = parser.parse(new_path)
       new_requirements = requirements.merge(visitor.route_translations)
-      [new_path, new_requirements]
+      [new_path, new_path_spec, new_requirements]
     end
 
     def translate_segment segment, locale
       segment = untokenize_segment segment
-      CGI.escape I18n.t segment, default: segment, scope: 'routes', locale: locale
+      I18n.t segment, default: segment, scope: 'routes', locale: locale
```
Comment (mdimas) [inline lib/i18n_routable/translation_assistant.rb]: just curious why the `CGI.escape` isn't necessary any more

File: lib/i18n_routable/config.rb (hunk 6 above)
Comment (RyanMcDonald) [inline lib/i18n_routable/config.rb]: http://stackoverflow.com/questions/20361428/rails-i18n-validation-deprecation-warning

The errors are raised whenever the following methods are called:
```I18n.config.default_locale=
I18n.config.locale=
I18n.translate
I18n.localize
I18n.transliterate
```
So basically, every test breaks.

> Before this change, if you passed an unsupported locale, Rails would silently switch to it if the locale is valid (i.e. if there is a corresponding locale file in the /config/locales folder), otherwise the locale would default to the config.i18n.default_locale configuration (which defaults to :en).

So we have to set it to false to resort to the old behaviour. Otherwise, we have to make sure I18n.available_locales is explicitly set. I tried to implement this, but it got a little messy with the way some of the tests work. For example, spec_routes.rb does `:locales => [{:gibberish => :gibb}, :aussie, :es, :fr, :'fr-CA']`, which, if we set as the available_locales, complains because {:gibberish => :gibb} can't be converted to a symbol ([source](https://github.com/svenfuchs/i18n/blob/f19893d84262261d9c8abe303f465ea28ba057e4/lib/i18n/config.rb#L55)).

Comment (mdimas) [inline lib/i18n_routable/config.rb]: I wonder if it might make more sense to just set up the tests to use real locales.

File: lib/i18n_routable/generating/localizable_url_helper.rb [hunk 8]
```diff
@@ -4,26 +4,25 @@ module LocalizableUrlHelper
 
       # we override this, but it really only removes positional_keys that are
       # locale or i18n_*
-      def define_url_helper_with_localize(route, name, kind, options)
```
Comment (mdimas) [inline lib/i18n_routable/generating/localizable_url_helper.rb]: I was playing around with this and removed this entire file and the tests still passed. All the extension was supposed to do is remove `:locale` and `:i18n_` segments, and it looks like `ActionDispatch::Routing::RouteSet::NamedRouteCollection::UrlHelper` might do that now already.

File: spec/i18n_routable/i18n_routable_spec.rb [hunk 9]
```diff
@@ -84,9 +84,9 @@ def resolve(path, http_verb=:get)
       posts_path(locale: :fr, :i18n_posts => 'puestos').should eql '/fr/messages'
     end
```
Comment (mdimas) [inline spec/i18n_routable/i18n_routable_spec.rb]: It would probably be good to have an additional test here that other params are appended correctly

```
    it 'should append any other params' do
      posts_path(locale: :fr, :i18n_posts => 'puestos', :foo => 'bar').should eql '/fr/messages?foo=bar'
    end
```

File: spec/i18n_routable/i18n_routable_spec.rb (hunk 4 above)
Comment (mdimas) [inline spec/i18n_routable/i18n_routable_spec.rb]: It doesn't seem like the `UrlHelper` module likes to just take a default arg that is used as the id params any more. Even after I removed the `localizable_url_helper` I got

```
post_path("post-title") => ''
``` 

however this test passes

```
    it 'should work without passing parameters in a hash' do
      post_path(id: "post-title").should eql "/posts/post-title"
    end
```

File: lib/i18n_routable/generating/localizable_url_helper.rb (hunk 8 above)
Comment (mdimas) [inline lib/i18n_routable/generating/localizable_url_helper.rb]: As far as I can tell it looks like `I18nRoutable::Outgoing::LocalizableFormatter` is doing everything that has to be done in terms of dealing with `:locale` and `:i18n_` params.

File: lib/i18n_routable/translation_assistant.rb (hunk 7 above)
Comment (RyanMcDonald) [inline lib/i18n_routable/translation_assistant.rb]: The updated Journey appears to already handle the escaping. So with the CGI.escape, it was double-escaping, which resulted in
```
expected: "/es/caf%C3%A9"
            got: "/es/caf%25C3%25A9"
```
[localizable_formatter:44](https://github.com/change/i18n_routable/blob/master/lib/i18n_routable/outgoing/localizable_formatter.rb#L44) calls `route.format`
[route.rb:68](https://github.com/rails/rails/blob/3e36db4406beea32772b1db1e9a16cc1e8aea14c/actionpack/lib/action_dispatch/journey/route.rb#L68) calls `@path_formatter.evaluate`
[visitors.rb:43](https://github.com/rails/rails/blob/3e36db4406beea32772b1db1e9a16cc1e8aea14c/actionpack/lib/action_dispatch/journey/visitors.rb#L43) which calls `param.escape`.

File: spec/i18n_routable/i18n_routable_spec.rb (hunk 4 above)
Comment (RyanMcDonald) [inline spec/i18n_routable/i18n_routable_spec.rb]: In that case, the test could be removed then? Since the purpose of it was to prove you don't have to pass a hash in... but now it looks like you do.

We could alternatively alter `localizable_url_helper` so that if they just pass in a string, we could build a hash with the `id` key referencing that string value. So it keeps it backwards compatible?

File: spec/i18n_routable/i18n_routable_spec.rb (hunk 9 above)
Comment (RyanMcDonald) [inline spec/i18n_routable/i18n_routable_spec.rb]: Added.

File: spec/i18n_routable/i18n_routable_spec.rb [hunk 10]
```diff
@@ -149,55 +153,55 @@ def resolve(path, http_verb=:get)
 
     it 'should understand the difference between display locales and backend locales' do
       I18n.locale = :gibb
-      new_post_url.should eql "http://www.example.com/gibberish/posts/the-new"
+      new_post_url.should eql "http://www.example.com/gibb/posts/the-new"
```
Comment (RyanMcDonald) [inline spec/i18n_routable/i18n_routable_spec.rb]: With enforce_available_locales set to true, we can no longer specify `{:gibberish => :gibb}` in the locales array, because setting available_locales calls `to_sym` on each item in the array. I changed it to just `gibb` now. Is this functionality that we need to support, or can we just leave it as `gibb`? I'm not exactly sure what display locales and backend locales means, but it looks like some of these tests might be useless without the previous functionality?

File: config/secrets.yml (hunk 1 above)
Comment (RyanMcDonald) [inline config/secrets.yml]: I figured out it was only requiring sessions because it was creating a TestApplication in spec/mocks.rb. I have removed the secrets.yml file now.

File: i18n_routable.gemspec (hunk 5 above)
Comment (RyanMcDonald) [inline i18n_routable.gemspec]: Where did you see that it's unsupported? I added it to I18n.available_locales and then set up this test which passed:
```
it "should accept 'en-US' :locale param as an option" do
      posts_path(:locale => :'en-US').should eql "/en-US/posts"
end
```

File: lib/i18n_routable.rb [hunk 11]
```diff
@@ -1,6 +1,5 @@
-require 'rails/all'
+require 'action_controller/railtie'
```
Comment (RyanMcDonald) [inline lib/i18n_routable.rb]: The reason I had to change this was because requiring rails/all would include ActiveRecord which would try to find database.yml. I ended up with these errors:
```
Failure/Error: Unable to find matching line from backtrace
     ActiveRecord::ConnectionNotEstablished:
       No connection pool for ActiveRecord::Base
```

File: spec/i18n_routable/i18n_routable_spec.rb (hunk 10 above)
Comment (erikogan) [inline spec/i18n_routable/i18n_routable_spec.rb]: For gibberish specifically, it is fine if we just make it gibb, since that is the locale we use for internal testing.

But I worry that we use that functionality (hash of display_locale => rails_locale) in our usage of i18n_routable, and this test was ensuring it worked (as a side-effect).

@mdimas Do you know if we still do that?

Comment (mdimas) [inline spec/i18n_routable/i18n_routable_spec.rb]: Monorail might still have some usage of display_locale. Or more likely there is still code that uses it, but it's not a used feature. So we might be able to remove it if so.

File: i18n_routable.gemspec [hunk 12]
```diff
@@ -18,9 +18,10 @@ Gem::Specification.new do |s|
...
 
   s.add_development_dependency "rspec", '~> 2.14.1'
   s.add_development_dependency "rspec-rails", '~> 2.14.1'
+  s.add_development_dependency 'debugger'
```
Comment (erikogan) [inline i18n_routable.gemspec]: This is going to cause trouble on MRI >= 2.0, and I suspect jRuby as well.

File: lib/i18n_routable/generating/localizable_url_helper.rb (hunk 8 above)
Comment (erikogan) [inline lib/i18n_routable/generating/localizable_url_helper.rb]: I would be all for removing the entire file, because I agree, it seems redundant.

File: lib/i18n_routable/outgoing/localizable_formatter.rb [hunk 13]
```diff
@@ -3,7 +3,7 @@ module I18nRoutable
   module Outgoing
     module LocalizableFormatter
 
-      def generate_with_localize key, name, options, recall = {}, parameterize = nil
+      def generate_with_localize name, options, recall = {}, parameterize = nil
```
Comment (erikogan) [inline lib/i18n_routable/outgoing/localizable_formatter.rb]: $#@#$% Thomas and his irrational hatred of parentheses.

File: lib/i18n_routable/translation_assistant.rb [hunk 14]
```diff
@@ -15,16 +15,18 @@ def untokenize_segment segment
...
     def convert_path_to_localized_regexp path, requirements={}, anchor= true
-      spec = Journey::Parser.new.parse(path)
+      parser = ActionDispatch::Journey::Parser.new
+      spec = parser.parse(path)
       visitor = RouteTranslationVisitor.new
       new_path = visitor.accept(spec)
+      new_path_spec = parser.parse(new_path)
       new_requirements = requirements.merge(visitor.route_translations)
-      [new_path, new_requirements]
+      [new_path, new_path_spec, new_requirements]
```
Comment (erikogan) [inline lib/i18n_routable/translation_assistant.rb]: Technically, changing this method signature is a breaking change to the API. While I am sure this is necessary, we should call it out in documentation.

File: lib/i18n_routable/translation_assistant.rb [hunk 15]
```diff
@@ -15,16 +15,18 @@ def untokenize_segment segment
 
     # /posts => ["/:i18n_posts", ["posts"]]
```
Comment (erikogan) [inline lib/i18n_routable/translation_assistant.rb]: Also, this comment is now incorrect.

File: spec/i18n_routable/i18n_routable_spec.rb (hunk 2 above)
Comment (erikogan) [inline spec/i18n_routable/i18n_routable_spec.rb]: Yes. It’s a shame that hash_for_* has been deprecated. But you can always call recognize_route on url_for_*, though that seems a bit back-assward if you ask me.

**Comment by erikogan** [comment]
Looks mostly OK to me. I agree with @mdimas: we should bump the version number for the breaking API changes (and moving support from Rails 3 => 4). We should call both of those out, as well as including a tag to make it easy to pull 0.4.x.

The only other change I would make is to remove the `debugger` development dependency, since that forces development on MRI 1.9.x.

File: i18n_routable.gemspec (hunk 5 above)
Comment (RyanMcDonald) [inline i18n_routable.gemspec]: Apparently en-US is unsupported, but en-us is? Is that an easy change in monorail?

**Comment by RyanMcDonald** [comment]
I've updated the readme to include some of the major changes to this version. Did I miss anything/is the information I added correctly stated?

**Comment by quaelin** [comment]
Cleaning up ancient PRs.  Please reopen if you care about this.
